Re-fetch more itineraries (part 1)

joonaojapalo opened the pull request:
Related issue: DT-5464

## Proposed Changes

  - Re-fetch more itineraries on using initial SummaryPage graphql query, if initial query responds with less than 5.
  - Re-fetch more itineraries when user request more earlier or later and response contains less than 3.
  - Refactor: SummaryPage.js to subdirectory
  - Refactor: graphql refetch logic move from SummaryPage component to higher-order-components.
  - Relay bundle middleware removed as currently providing only marginal performance improvement and som issues with pageCursor queries.
  - Partial result rendering implementation will be splitted to another PR

## Pull Request Check List

  - A reasonable set of unit tests is included
  - Console does not show new warnings/errors
  - Changes are documented or they are self explanatory
  - This pull request does not have any merge conflicts
  - All existing tests pass in CI build
  - Code coverage does not decrease (unless measured incorrectly)

## Review

  - Read and verify the code changes
  - Test the functionality by running the UI locally with all popular browsers available in your platform
  - Check that the implementation matches the design, when such one is defined in a Jira issue
  - Merge the pull request

optionsome commented on app/component/SummaryPage/hoc/withFetchMoreItineraries.js:
```diff
@@ -0,0 +1,312 @@
+          .slice() // Need to copy because result is readonly
+          .reverse()
+          .filter(
+            itinerary => !itinerary.legs.every(leg => leg.mode === 'WALK'),
```
I think we could actually leave the WALK mode out of the public transportation requests because OTP2 returns walk legs for public transportation even without it. Then we wouldn't need to do this filtering in UI which could simplify the logic a bit.

optionsome replied:
We won't do this because the WALK only itineraries are now sometimes rendered.

optionsome commented on app/component/SummaryPage/hooks/useAutofetchQuery.js:
```diff
@@ -0,0 +1,238 @@
+    error: undefined,
+  };
+
+  const ACTION_DATA = 'data';
+  const ACTION_START = 'start';
+  const ACTION_ERROR = 'error';
+  const ACTION_DATA_COMPLETE = 'data-complete';
+
+  const STATUS_PROCESSING = 'PROCESSING';
+  const STATUS_COMPLETE = 'COMPLETE';
+  const STATUS_FETCHING = 'FETCHING';
+  const STATUS_READY = 'READY';
```
Should we change these to be in the enum object format and maybe in some different file we can, for example, check that the status is something like `status === AutofetchQueryStatuses.Complete`?

optionsome commented on app/component/SummaryPage/hooks/useAutofetchRelay.js:
```diff
@@ -0,0 +1,213 @@
+import getUniqItineraries from '../utils/uniqItineraries';
+import createCancelable from '../utils/createCancelable';
+
+const ACTION_MOUNTED = 'MOUNTED';
+const ACTION_DATA_UPDATE = 'DATA_UPDATE';
+const ACTION_QUERY_UPDATE = 'QUERY_UPDATE';
+const ACTION_ERROR = 'ERROR';
+
+const STATUS_FETCHING = 'FETCHING';
+const STATUS_REFETCHING = 'REFETCHING';
+const STATUS_COMPLETE = 'COMPLETE';
```
Same here as above.

optionsome reviewed (commented):
Tested this quickly and found a couple of problems. Changing something from settings causes no new normal public transportation plan query to be sent. If the search finds no itineraries because no itineraries can be found within the search window, the next page query seems to be canceled.

vesameskanen committed:
feat: update schema, add initial planConnection query

vesameskanen committed:
fix: use relayShape, remove unused headers from context

vesameskanen committed:
fix: don't use eslint in external code

vesameskanen committed:
feat: test the new planConnection in ItineraryPage

vesameskanen committed:
Merge branch 'fix-all-proptypes' into DT-5464

vesameskanen committed:
chore: remove unused store and prop

vesameskanen committed:
chore: try removing redundant graphql layer

vesameskanen committed:
chore: try reducing graphql in middle container level

vesameskanen committed:
chore: simplify more

vesameskanen committed:
chore: remove unused fields from top level itinerary queries

vesameskanen committed:
feat: add graphql fragment to itinerary

vesameskanen committed:
fix: remove redundant prop

vesameskanen committed:
fix: clean up ItineraryList graphql

vesameskanen committed:
fix: clean up itinerary graphql

vesameskanen committed:
chore: tune itinerary fragments

vesameskanen committed:
Merge branch 'v3' into DT-5464

vesameskanen committed:
fix: simplify settings handling

vesameskanen committed:
feat: move mode setup to planParamUtil

vesameskanen committed:
feat: compose planConnection modes for all itinerary types

vesameskanen committed:
chore: plan and itinerary shapes which match planConnection

vesameskanen committed:
fix: throw away old plan  queries, only one planConnection query is enough

vesameskanen committed:
chore: rename itinerary query file

vesameskanen committed:
feat: refactor planParamUtils for planConnection

Add new util function which implement consistent way to check if
particular query type is needed.

vesameskanen committed:
chore: bike mode filter now includes only transit

vesameskanen committed:
feat: use new plan and itinerary shapes

vesameskanen committed:
feat: add planEdgeShape

vesameskanen committed:
feat: use planEdgeShape

vesameskanen committed:
feat: split and update all itinerary page queries

vesameskanen committed:
chore: update and disable temporarily some tests

vesameskanen committed:
fix: never modify a prop during rendering!

Relay objects are read-only, which reveals old bugs

vesameskanen committed:
fix: lots of new bugs due to extensive amount of code changes

vesameskanen committed:
fix: don't search transit if transitModes is empty

vesameskanen committed:
fix: time value handling

vesameskanen committed:
fix: update planParamUtil tests

vesameskanen committed:
feat: add paging params into PlanConnection

vesameskanen committed:
feat: iterate itinerary search using page cursors

vesameskanen committed:
fix: correct direct mode

vesameskanen committed:
feat: use paging in earlier/later searches

vesameskanen committed:
fix: proper paging logig for earlier/later
